src/xeroClient.ts

```diff
@@ -10,112 +10,167 @@ export interface IXeroClientConfig {
   scopes: string[],
   state?: string
 }
+export interface XeroIdToken {
+  nbf: number
+  exp: number
+  iss: string,
+  aud: string
+  iat: number
+  at_hash: string
+  sid: string
+  sub: string
+  auth_time: number
+  idp: string
+  xero_userid: string
+  global_session_id: string
+  preferred_username: string
+  email: string
+  given_name: string
+  family_name: string
+  amr: string[]
+}
 
-export class XeroClient {
+export interface XeroAccessToken {
+  nbf: number
+  exp: number
+  iss: string
+  aud: string
+  client_id: string
+  sub: string
+  auth_time: number
+  idp: string
+  xero_userid: string
+  global_session_id: string
+  jti: string
+  scope: string[]
+  amr: string[]
+}
 
+export class XeroClient {
   constructor(private readonly config: IXeroClientConfig) {
-    // need to set access token before use
     this.accountingApi = new xero.AccountingApi(); 
-    this.buildClient()
   }
 
   private tokenSet: TokenSet = new TokenSet
-  private _tenantIds: string[] = ['']
+  private _tenants: any[] = []
 
   readonly accountingApi: xero.AccountingApi;
 
-  private openIdClient: any; // from openid-client
+  openIdClient: any; // from openid-client
   
-  get tenantIds(): string[] {
-    return this._tenantIds;
+  get tenants(): any[] {
+    return this._tenants;
   }
-
-  async buildClient() {
+  
+  async initialize() {
     const issuer = await Issuer.discover('https://identity.xero.com');
     this.openIdClient = new issuer.Client({
       client_id: this.config.clientId,
       client_secret: this.config.clientSecret,
       redirect_uris: this.config.redirectUris,
     });
     this.openIdClient[custom.clock_tolerance] = 5
+    return this
   }
 
   async buildConsentUrl() {
+    await this.initialize()
     const url = this.openIdClient.authorizationUrl({
       redirect_uri: this.config.redirectUris[0],
       scope: this.config.scopes.join(' ') || 'openid email profile'
     });
     return url;
   }
 
-  async setAccessTokenFromRedirectUri(url: string) {
+  async apiCallback(url: string): Promise<TokenSet> {
     const params = this.openIdClient.callbackParams(url)
     const check = {...params}
     this.tokenSet = await this.openIdClient.callback(this.config.redirectUris[0], params, check);
-    this.setAccessTokenForAllApis();
+    this.setAccessToken();
+    return this.tokenSet
+  }
 
-    await this.fetchConnectedTenantIds();
+  async disconnect(tenantId: string): Promise<TokenSet> {
+    await this.queryApi('DELETE', `https://api.xero.com/connections/${tenantId}`)
```

Comment:
pretty sure this takes connectionID not tenantID, we should double check prior to merging PR

Reply:
@RettBehrens is correct the Connection object returns both an "id" and "tenantId".

It's the "id" you need to pass.